src/RuleSet/RuleSets.php

```diff
@@ -67,4 +68,45 @@ public static function getSetDefinition(string $name): RuleSetDescriptionInterfa
 
         return $definitions[$name];
     }
+
+    /**
+     * @param class-string<RuleSetDescriptionInterface> $class
+     */
+    public static function registerRuleSet(string $name, string $class): void
```

Comment:
# clarification needed:

if it's already there somewhere, sorry for that - please fingerprint the place.

let's imagine that I want to enable the custom ruleset `foo/@bar` (that includes rule `foo/baz`).
As user of `foo/@bar`, i have no clue that rule `foo/baz` exists in first place. Do I need to call `registerCustomFixers()` for `foo/baz`?

I believe we should not and that we should have test for it

Reply:
(and similar, custom ruleset may include another custom ruleset)

Reply:
I believe it is not covered, BUT it shouldn't be a blocker here. We can cover simpler scenario and provide support for auto-discovery later. I don't even know if it would be possible to handle nested custom sets only based on rules within set, because we need FQCN of ruleset and only name is available in the rules (when nesting). Please don't kill this PR with such complex requirement 🙏.

Reply:
Auto-discovery has not been implemented and I honestly didn't even consider it before. Currently you would have to register all ruleset manually – and I haven't even tested this, so I don't know if it would work. I would expect it to work, but...

Part of the problem here is that currently the plan is to allow to freely name the rulesets. Maybe, in the future, we could implement a rule for registering rulesets from the rules, too? This could possibly again mean overlapping names.

Alternative option is to promote extending the other custom rulesets when there's a need for that. This would allow to inherit and add/remove rules based on different needs.

Reply:
I believe releasing support for ruleset without autoregistration of ruleset content will be a huge shortcoming of the feature itself. And it's a good challenge to see how usage ready is the custom ruleset registration by covering this path.

That's exactly why I brought this concern. Currently, for registering custom Fixer, we provide instance of it to the Config::registerCustomFixer(). But for ruleset, this PR proposes to provide only it's name/fqcn. lack of consistency is causing problems.

Reply:
It wouldn't change anything if we change `class-string<RuleSetInterface>` to `RuleSetInterface` and require instance of ruleset - problem remains because when resolving config you don't have FQCNs for nested sets, only their names.

To make auto-discovery possible, `RuleSetInterface` should define something like `getDependentRuleSets(): array<string, class-string<RuleSetInterface>>` but then we're again introducing BC break.

I believe that registering custom rule set that includes another custom ruleset within its rules is so edge case scenario, that it shouldn't block this PR. At this point we could just print warning/error to the user that rules' configuration contains reference to unknown ruleset and it should be registered in order to work properly.